stream_alert/alert_processor/outputs/credentials/provider.py

```diff
@@ -0,0 +1,654 @@
+
+    def load_credentials(self, descriptor):
+        local_cred_location = self.get_file_path(descriptor)
+        with open(local_cred_location, 'rb') as cred_file:
+            encrypted_credentials = cred_file.read()
+
+        return Credentials(encrypted_credentials, True, self._region)
+
+    def has_credentials(self, descriptor):
+        return os.path.exists(self.get_file_path(descriptor))
+
+    def save_credentials(self, descriptor, credentials):
+        if not credentials.is_encrypted():
+            LOGGER.error('Error: Writing unencrypted credentials to disk is disallowed.')
+            return False
+
+        with self.offer_fileobj(descriptor) as file_handle:
+            file_handle.write(credentials.data())
+        return True
+
+    @staticmethod
+    def clear():
+        """Removes the local secrets directory that may be left from previous runs"""
+        secrets_dirtemp_dir = LocalFileDriver.get_local_credentials_temp_dir()
+
+        # Check if the folder exists, and remove it if it does
+        if os.path.isdir(secrets_dirtemp_dir):
+            shutil.rmtree(secrets_dirtemp_dir)
+
+    def offer_fileobj(self, descriptor):
+        """Opens a file-like object and returns it.
+
+        If you use the return value in a `with` statement block then the file descriptor
+        will auto-close.
+
+        Return:
+            file object
+        """
+        file_path = self.get_file_path(descriptor)
+        if not os.path.exists(file_path):
+            os.makedirs(os.path.dirname(file_path))
+
+        return open(file_path, 'a+b')  # read+write and in binary mode
+
+    def get_file_path(self, descriptor):
+        local_cred_location = os.path.join(
+            self._temp_dir,
+            get_formatted_output_credentials_name(self._service_name, descriptor)
+        )
+        return local_cred_location
+
+    @staticmethod
+    def get_local_credentials_temp_dir():
+        """Returns a temporary directory on the filesystem to store encrypted credentials.
+
+        Will automatically create the new directory if it does not exist.
+
+        Returns:
+            str: local path for stream_alert_secrets tmp directory
+        """
+        temp_dir = os.path.join(tempfile.gettempdir(), "stream_alert_secrets")
+
+        # Check if this item exists as a file, and remove it if it does
+        if os.path.isfile(temp_dir):
+            os.remove(temp_dir)
+
+        # Create the folder on disk to store the credentials temporarily
+        if not os.path.exists(temp_dir):
+            os.makedirs(temp_dir)
+
+        return temp_dir
+
+
+class SpooledTempfileDriver(CredentialsProvidingDriver, FileDescriptorProvider):
+    """Driver for fetching credentials that are stored in memory in file-like objects."""
+
+    SERVICE_SPOOLS = {}
+
+    def __init__(self, service_name, region):
+        self._service_name = service_name
+        self._region = region
+        self._spools = {}
```

Comment:
I think this may be a leftover/unused instance attribute.. potentially left after moving to a class property?